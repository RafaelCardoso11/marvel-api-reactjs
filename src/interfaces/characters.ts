import { ICharacter } from "./character";

export interface ICharacters {
  data: {
    results: ICharacter[];
  };
}
