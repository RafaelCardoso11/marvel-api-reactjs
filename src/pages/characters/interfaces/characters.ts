export interface ICharacters {
  id: number;
  name: string;
  thumbnail: { path: string; extension: string };
}
