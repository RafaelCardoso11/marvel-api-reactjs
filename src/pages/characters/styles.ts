import { styled } from "styled-components";

const CharactersContainer = styled.div`
  margin: 0 5vw;
  height: 100vh;
  padding: 2px 10px;
`;

const CharactersHeaderContainer = styled.header`
  padding: 20px 2rem;
`;

const CharactersNotCharactersContainer = styled.div``;
const CharactersCardsContainer = styled.div``;
export {
  CharactersContainer,
  CharactersHeaderContainer,
  CharactersNotCharactersContainer,
  CharactersCardsContainer,
};
