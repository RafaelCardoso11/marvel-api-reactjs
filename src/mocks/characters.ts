import { ICharacters } from "@/interfaces/characters";

export const charactersMock: ICharacters = {
  data: {
    results: [
      {
        id: 1011334,
        name: "3-D Man",
        thumbnail: {
          path: "http://i.annihil.us/u/prod/marvel/i/mg/c/e0/535fecbbb9784",
          extension: "jpg",
        },
        description: "lorem ipsum dolor sit amet, consectetur adip"
      },
      {
        id: 1011335,
        name: "Hulk",
        thumbnail: {
          path: "http://i.annihil.us/u/prod/marvel/i/mg/c/e0/535fecbbb9784",
          extension: "jpg",
        },
        description: "lorem ipsum dolor sit amet, consectetur adip 2"
      },
    ],
  }
};
