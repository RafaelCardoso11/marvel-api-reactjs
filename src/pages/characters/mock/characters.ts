import { ICardsSearchProps } from "../interfaces/charactersSearchProps";

export const characters: ICardsSearchProps = {
  data: {
    results: [
      {
        id: 1011334,
        name: "3-D Man",
        thumbnail: {
          path: "http://i.annihil.us/u/prod/marvel/i/mg/c/e0/535fecbbb9784",
          extension: "jpg",
        },
      },
    ],
  },
};
