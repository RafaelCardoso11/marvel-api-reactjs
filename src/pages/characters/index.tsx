import { useDataFetcher } from "@/hooks/useDataFetcher";
import { useState } from "react";

import { ICardsSearchProps } from "./interfaces/charactersSearchProps";
import { Cards } from "./components/cards";
import { StyledContainer, StyledContainerSearch } from "./styles";

export const Characters = () => {
  const [queryParams, setQueryParams] = useState("characters");
  const [searchCharacters, setSearchCharacters] = useState<string>("");
  const { data: characters } = useDataFetcher<ICardsSearchProps>(queryParams);

  const handleSearch = async (event: React.ChangeEvent<HTMLInputElement>) => {
    const searchCharacters = event.target.value;
    setSearchCharacters(searchCharacters);
    if (searchCharacters) {
      setQueryParams("characters?nameStartsWith=" + searchCharacters);
    } else {
      setQueryParams("characters");
    }
  };

  return (
    <StyledContainer data-testid="id-container-characters">
      <StyledContainerSearch data-testid="id-container-search-characters">
        <input
          type="text"
          placeholder="Pesquisar..."
          data-testid="id-searchCharacters"
          value={searchCharacters}
          onChange={handleSearch}
        />
      </StyledContainerSearch>
      {characters && <Cards data={characters.data} data-testid="id-cards-characters"/>}
    </StyledContainer>
  );
};
