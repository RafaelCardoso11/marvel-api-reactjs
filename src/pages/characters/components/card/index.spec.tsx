import { cleanup, fireEvent, render } from "@testing-library/react";
import { Card } from ".";
import { charactersMock } from "@/mocks/characters";
import RoutersContextTesting from "@/tests/routers";
import { Route } from "react-router-dom";

describe("<Cards/>", () => {
  afterEach(cleanup);
  afterAll(() => {
    jest.unmock("@/hooks/useDataFetcher");
  });
  beforeEach(() => {
    jest.clearAllMocks();
  });
  it("Deveria renderizar na tela um container card, texto do name e imagem do card", () => {
    const screen = render(
      <RoutersContextTesting>
        <Card character={charactersMock.data.results[0]} />
      </RoutersContextTesting>
    );
    const cardsContainer = screen.getByTestId("id-card-container");
    const typographNameCard = screen.getByTestId("id-typograph-name-content");
    const imgCardCharacter = screen.getByTestId("id-img-content");

    expect(cardsContainer).toBeInTheDocument();
    expect(typographNameCard).toBeInTheDocument();
    expect(imgCardCharacter).toBeInTheDocument();
    expect(screen.container).toMatchSnapshot();
    expect.assertions(4);
  });

  it("Deveria renderizar a imagem e nome do personagem ao injeta-lo por props", () => {
    const oneCharactersForCharactersMock = charactersMock.data.results[0];
    const screen = render(
      <RoutersContextTesting>
        <Card character={oneCharactersForCharactersMock} />
      </RoutersContextTesting>
    );
    const cardsContainer = screen.getByTestId("id-card-container");
    const typographNameCard = screen.getByTestId("id-typograph-name-content");
    const imgCardCharacter = screen.getByTestId("id-img-content");

    expect(cardsContainer).toBeInTheDocument();
    expect(typographNameCard).toBeInTheDocument();
    expect(imgCardCharacter).toBeInTheDocument();

    const {
      name,
      thumbnail: { extension, path },
    } = oneCharactersForCharactersMock;
    const srcCharacter = `${path}.${extension}`;
    const altCharacter = `Imagem do personagem ${name}`;

    expect(typographNameCard).toHaveTextContent(name);

    expect(imgCardCharacter).toHaveAttribute("src", srcCharacter);
    expect(imgCardCharacter).toHaveAttribute("alt", altCharacter);

    expect(screen.container).toMatchSnapshot();
    expect.assertions(7);
  });
  it("Deveria redirecionar para a página de informações do personagem ao clicar no card ", () => {
    const textInPageInfo = "Página de informações";
    const oneCharactersForCharactersMock = charactersMock.data.results[0];
    const { id } = oneCharactersForCharactersMock;

    const screen = render(
      <RoutersContextTesting
        routes={[
          <Route
            path={`/characters/${id}`}
            element={<>{textInPageInfo}</>}
            key={id}
          />,
        ]}
      >
        <Card character={oneCharactersForCharactersMock} />
      </RoutersContextTesting>
    );

    const card = screen.getByTestId("id-card-container");

    fireEvent.click(card);

    
    expect(window.location.pathname).toBe(`/characters/${id}`);
    
    const pageInfo = screen.getByText(textInPageInfo) 
    expect(window.location.pathname).toBe(`/characters/${id}`);
    expect(pageInfo).toBeInTheDocument()
    expect.assertions(3);
  });
});
