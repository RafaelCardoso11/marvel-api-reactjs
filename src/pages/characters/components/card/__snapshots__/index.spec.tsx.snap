// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Cards/> Deveria renderizar a imagem e nome do personagem ao injeta-lo por props 1`] = `
<div>
  <div
    class="sc-kvAqsT efdGIg"
    data-testid="id-card-container"
  >
    <h2
      class="sc-cdtKKh hnyVce"
      data-testid="id-typograph-name-content"
    >
      3-D Man
    </h2>
    <img
      alt="Imagem do personagem 3-D Man"
      class="sc-grJzXs cqACap"
      data-testid="id-img-content"
      src="http://i.annihil.us/u/prod/marvel/i/mg/c/e0/535fecbbb9784.jpg"
    />
  </div>
</div>
`;

exports[`<Cards/> Deveria renderizar na tela um container card, texto do name e imagem do card 1`] = `
<div>
  <div
    class="sc-kvAqsT efdGIg"
    data-testid="id-card-container"
  >
    <h2
      class="sc-cdtKKh hnyVce"
      data-testid="id-typograph-name-content"
    >
      3-D Man
    </h2>
    <img
      alt="Imagem do personagem 3-D Man"
      class="sc-grJzXs cqACap"
      data-testid="id-img-content"
      src="http://i.annihil.us/u/prod/marvel/i/mg/c/e0/535fecbbb9784.jpg"
    />
  </div>
</div>
`;
